translator/opencensus/oc_to_internal.go

```diff
@@ -0,0 +1,405 @@
+				if ok && strVal != nil {
+					var otlpKind data.SpanKind
+					switch tracetranslator.OpenTracingSpanKind(truncableStringToStr(strVal.StringValue)) {
+					case tracetranslator.OpenTracingSpanKindConsumer:
+						otlpKind = data.SpanKindCONSUMER
+					case tracetranslator.OpenTracingSpanKindProducer:
+						otlpKind = data.SpanKindPRODUCER
+					default:
+						return data.SpanKindUNSPECIFIED
+					}
+					delete(ocAttrs.AttributeMap, tracetranslator.TagSpanKind)
+					return otlpKind
+				}
+			}
+		}
+		return data.SpanKindUNSPECIFIED
+
+	default:
+		return data.SpanKindUNSPECIFIED
+	}
+}
+
+func ocEventsToInternal(ocEvents *octrace.Span_TimeEvents) (ptrs []*data.SpanEvent, droppedCount uint32) {
+	if ocEvents == nil {
+		return
+	}
+
+	droppedCount = uint32(ocEvents.DroppedMessageEventsCount + ocEvents.DroppedAnnotationsCount)
+
+	evenCount := len(ocEvents.TimeEvent)
+	if evenCount == 0 {
+		return
+	}
+
+	// Create slice that holds pointers and another slice that holds structs at once
+	// to avoid individual allocations of structs.
+	ptrs = make([]*data.SpanEvent, 0, evenCount)
+	content := data.NewSpanEventSlice(evenCount)
+
+	for i, ocEvent := range ocEvents.TimeEvent {
+		if ocEvent == nil {
+			continue
+		}
+
+		// Point one element in slice of pointers to an element in slice of structs.
+		event := &content[i]
+		ptrs = append(ptrs, event)
+
+		event.SetTimestamp(timestampToUnixnano(ocEvent.Time))
+
+		switch teValue := ocEvent.Value.(type) {
+		case *octrace.Span_TimeEvent_Annotation_:
+			if teValue.Annotation != nil {
+				event.SetName(truncableStringToStr(teValue.Annotation.Description))
+				attrs := ocAttrsToInternal(teValue.Annotation.Attributes)
+				event.SetAttributes(attrs)
+			}
+
+		case *octrace.Span_TimeEvent_MessageEvent_:
+			event.SetAttributes(ocMessageEventToInternalAttrs(teValue.MessageEvent))
+
+		default:
+			event.SetName("An unknown OpenCensus TimeEvent type was detected when translating")
+		}
+	}
+	return
+}
+
+func ocLinksToInternal(ocLinks *octrace.Span_Links) (ptrs []*data.SpanLink, droppedCount uint32) {
+	if ocLinks == nil {
+		return
+	}
+
+	droppedCount = uint32(ocLinks.DroppedLinksCount)
+
+	linkCount := len(ocLinks.Link)
+	if linkCount == 0 {
+		return
+	}
+
+	// Create slice that holds pointers and another slice that holds structs at once
+	// to avoid individual allocations of structs.
+	ptrs = make([]*data.SpanLink, 0, linkCount)
+	content := data.NewSpanLinkSlice(linkCount)
+
+	for i, ocLink := range ocLinks.Link {
+		if ocLink == nil {
+			continue
+		}
+
+		// Point one element in slice of pointers to an element in slice of structs.
+		link := &content[i]
+		ptrs = append(ptrs, link)
+
+		attrs := ocAttrsToInternal(ocLink.Attributes)
+		link.SetTraceID(data.TraceIDFromBytes(ocLink.TraceId))
+		link.SetSpanID(data.SpanIDFromBytes(ocLink.SpanId))
+		link.SetTraceState(ocTraceStateToInternal(ocLink.Tracestate))
+		link.SetAttributes(attrs)
+	}
+	return ptrs, droppedCount
```

Comment:
You're mixing naked and normal returns in this function.
I personally dislike naked returns, but should be consistent at least IMO.

Reply:
fixed